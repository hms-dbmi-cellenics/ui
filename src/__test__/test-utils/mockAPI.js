import _ from 'lodash';

// Default platform data
import cellSetsData from '__test__/data/cell_sets.json';
import backendStatusData from '__test__/data/backend_status.json';
import processingConfigData from '__test__/data/processing_config.json';
import mockDemoExperiments from '__test__/test-utils/mockData/mockDemoExperiments.json';

import fake from '__test__/test-utils/constants';

import {
  responseData,
} from '__test__/test-utils/mockData';

const promiseResponse = (
  response,
  options = {},
) => Promise.resolve(new Response(response, options));

const statusResponse = (code, body) => (
  Promise.resolve({
    status: code,
    body: JSON.stringify(body),
  })
);

const delayedResponse = (response, delay = 10000, options = {}) => new Promise((resolve) => {
  setTimeout(() => {
    resolve(new Response(response, options));
  }, delay);
});

const workerDataResult = (data) => Promise.resolve(_.cloneDeep(data));

const fetchWorkMock = (
  mockedResults,
) => ((experimentId, body) => {
  if (typeof mockedResults[body.name] === 'function') {
    return mockedResults[body.name]();
  }

  return workerDataResult(mockedResults[experimentId]);
});

const generateDefaultMockAPIResponses = (experimentId) => ({
  [`experiments/${experimentId}$`]: () => promiseResponse(
    JSON.stringify(responseData.experiments.find(({ id }) => id === experimentId)),
  ),
  [`experiments/${experimentId}/processingConfig$`]: () => promiseResponse(
    JSON.stringify(processingConfigData),
  ),
  [`experiments/${experimentId}/cellSets$`]: () => promiseResponse(
    JSON.stringify(cellSetsData),
  ),
  [`experiments/${experimentId}/backendStatus$`]: () => promiseResponse(
    JSON.stringify(backendStatusData),
  ),
  experiments$: () => promiseResponse(
    JSON.stringify(responseData.experiments),
  ),
  [`experiments/${experimentId}/samples$`]: () => promiseResponse(
    JSON.stringify(responseData.samples[0]),
  ),
  '/v2/experiments/examples$': () => promiseResponse(
    JSON.stringify(mockDemoExperiments),
  ),
  'experiments/clone$': () => promiseResponse(
    JSON.stringify(fake.CLONED_EXPERIMENT_ID),
  ),
  [`/v2/workRequest/${experimentId}`]: () => statusResponse(200, 'OK'),
});

const mockAPI = (apiMapping) => (req) => {
  const path = req.url;

  const key = _.find(Object.keys(apiMapping), (matcherStr) => new RegExp(matcherStr).test(path));

  if (!key) {
    return statusResponse({
      status: 404,
      body: `Path ${path} is undefined`,
    });
  }

  return apiMapping[key](req);
};

export default mockAPI;
export {
  generateDefaultMockAPIResponses,
  promiseResponse,
  statusResponse,
  delayedResponse,
  workerDataResult,
  fetchWorkMock,
};
