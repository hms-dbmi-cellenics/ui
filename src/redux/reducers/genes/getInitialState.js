import ExpressionMatrix from 'utils/ExpressionMatrix/ExpressionMatrix';

const initialViewState = {
  fetching: false,
  error: false,
  data: [],
};

const getInitialState = () => ({
  properties: {
    loading: [],
    views: {},
    data: {},
  },
  expression: {
    views: {},
    full: {
      loading: [],
      error: false,
      matrix: new ExpressionMatrix(),
    },
    downsampled: {
      loading: [],
      error: false,
      matrix: new ExpressionMatrix(),
      cellOrder: [],
      ETag: null,
    },
  },
  selected: [],
  focused: undefined,
  markers: {
    loading: false,
    error: false,
    ETag: null,
  },
});

export { initialViewState };
export default getInitialState;
