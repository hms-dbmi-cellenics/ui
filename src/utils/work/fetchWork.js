import { Environment, isBrowser } from 'utils/deploymentInfo';

import cache from 'utils/cache';
import dispatchWorkRequest from 'utils/work/dispatchWorkRequest';
import downloadFromS3 from 'utils/work/downloadFromS3';
import waitForWorkRequest from 'utils/work/waitForWorkRequest';

// retrieveData will try to get the data for the given experimentId and ETag from
// the fastest source possible. It will try to get the data in order from:
// 1. Browser cache
// 2. S3 (cache) via signedURL
// 3. Worker
//   3.1. Via socket (small data)
//   3.2. Via S3 (large data) via signedURL
const retrieveData = async (experimentId,
  ETag,
  signedUrl,
  request,
  timeout,
  body,
  dispatch,
  useCache) => {
  // 1. Check if we have the ETag in the browser cache (no worker)
  const cachedData = useCache ? await cache.get(ETag) : null;
  if (cachedData) {
    return cachedData;
  }

  // 2. Check if data is cached in S3 so we can download from the signed URL (no worker)
  if (signedUrl) {
    return await downloadFromS3(body.name, signedUrl);
  }

  // 3. If we don't have signedURL, wait for the worker to send us the data via
  // - the data via socket
  // - the signedURL to download the data from S3
  const { signedUrl: workerSignedUrl, data } = await waitForWorkRequest(
    ETag,
    experimentId,
    request,
    timeout,
    dispatch,
  );

  // 3.1. The worker send the data via socket because it's small enough
  if (data) {
    return data;
  }
  // 3.2. The worker send a signedUrl to download the data
  return await downloadFromS3(body.name, workerSignedUrl);
};

const fetchWork = async (
  experimentId,
  body,
  getState,
  dispatch,
  optionals = {},
) => {
  const {
    extras = undefined,
    timeout = 180,
    broadcast = false,
    onETagGenerated = () => { },
  } = optionals;

  if (!isBrowser) {
    throw new Error('Disabling network interaction on server');
  }

  const { environment } = getState().networkResources;
  if (environment === Environment.DEVELOPMENT && !localStorage.getItem('disableCache')) {
    localStorage.setItem('disableCache', 'true');
  }

  // 1. Contact the API to get ETag and possible S3 signedURL
  const { ETag, signedUrl, request } = await dispatchWorkRequest(
    experimentId,
    body,
    timeout,
    {
      broadcast,
      ...extras,
    },
    dispatch,
  );

  onETagGenerated(ETag);

  const useCache = body.name !== 'GeneExpression';

  // 2. Try to get the data from the fastest source possible
  const data = await retrieveData(
    experimentId, ETag, signedUrl, request, timeout, body, dispatch, useCache,
  );

  // 3. Cache the data in the browser
  if (useCache) {
    await cache.set(ETag, data);
  }

  return data;
};

export default fetchWork;
