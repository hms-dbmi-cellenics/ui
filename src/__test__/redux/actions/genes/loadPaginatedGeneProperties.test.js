import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import fetchMock, { enableFetchMocks } from 'jest-fetch-mock';

import initialStateExperimentSettings from 'redux/reducers/experimentSettings/initialState';
import loadPaginatedGeneProperties from 'redux/actions/genes/loadPaginatedGeneProperties';
import getInitialState from 'redux/reducers/genes/getInitialState';
import {
  GENES_PROPERTIES_LOADING,
  GENES_PROPERTIES_LOADED_PAGINATED,
  GENES_PROPERTIES_ERROR,
} from 'redux/actionTypes/genes';

import fetchWork from 'utils/work/fetchWork';

import mockAPI, { generateDefaultMockAPIResponses, workerDataResult } from '__test__/test-utils/mockAPI';
import processingConfigData from '__test__/data/processing_config.json';

jest.mock('utils/work/fetchWork');

jest.mock('utils/getTimeoutForWorkerTask', () => ({
  __esModule: true, // this property makes it work
  default: () => 60,
}));

const mockStore = configureStore([thunk]);

const experimentId = '1234';
const properties = ['a', 'b', 'c'];
const componentUuid = 'asd';

const backendStatus = {
  [experimentId]: {
    status: {
      pipeline: {
        status: 'SUCCEEDED',
        startDate: '2021-01-01T01:01:01.000Z',
      },
    },
  },
};

describe('loadPaginatedGeneProperties action', () => {
  beforeEach(async () => {
    jest.clearAllMocks();

    fetchWork
      .mockReset()
      .mockImplementation(() => null);

    enableFetchMocks();
    fetchMock.resetMocks();
    fetchMock.doMock();
    fetchMock.mockIf(/.*/, mockAPI(generateDefaultMockAPIResponses(experimentId)));
  });

  it('Does not dispatch when some of the properties are already loading', async () => {
    const initialState = getInitialState();
    const store = mockStore({
      backendStatus,
      genes:
      {
        ...initialState,
        properties: {
          ...initialState.properties,
          loading: ['b'],
        },
      },
      experimentSettings: initialStateExperimentSettings,
    });

    store.dispatch(loadPaginatedGeneProperties(experimentId, properties, componentUuid, {}));
    expect(store.getActions().length).toEqual(0);
  });

  it('Dispatches appropriately on success condition', async () => {
    const initialState = getInitialState();
    const store = mockStore({
      genes: { ...initialState },
      backendStatus,
      networkResources: {
        environment: 'testing',
      },
      experimentSettings: {
        ...initialStateExperimentSettings,
        ...{ processing: processingConfigData.processingConfig },
      },
    });

    fetchWork
      .mockReset()
      .mockImplementation(() => workerDataResult({
        total: 2,
        gene_names: ['a', 'b'],
        dispersions: [1, 1],
      }));

    const tableState = {
      sorter: {
        field: 'gene_names',
        order: 'ascend',
      },
      pagination: {
        current: 2,
        pageSize: 20,
      },
    };

    await store.dispatch(
      loadPaginatedGeneProperties(experimentId, properties, componentUuid, tableState),
    );

    expect(store.getActions().length).toEqual(2);

    const loadingAction = store.getActions()[0];
    expect(loadingAction).toMatchSnapshot();
    expect(loadingAction.type).toEqual(GENES_PROPERTIES_LOADING);

    const loadedAction = store.getActions()[1];
    expect(loadedAction).toMatchSnapshot();
    expect(loadedAction.type).toEqual(GENES_PROPERTIES_LOADED_PAGINATED);

    const dispatchParams = fetchWork.mock.calls[0];
    expect(dispatchParams).toMatchSnapshot();
  });

  it('Dispatches appropriately on error condition', async () => {
    const store = mockStore({
      genes: { ...getInitialState() },
      backendStatus,
      networkResources: {
        environment: 'testing',
      },
      experimentSettings: {
        ...initialStateExperimentSettings,
        ...{ processing: processingConfigData.processingConfig },
      },
    });

    fetchWork
      .mockReset()
      .mockImplementation(() => Promise.reject(new Error('random error!')));

    const tableState = {
      sorter: {
        field: 'gene_names',
        order: 'ascend',
      },
      pagination: {
        current: 2,
        pageSize: 20,
      },
    };

    await store.dispatch(
      loadPaginatedGeneProperties(experimentId, properties, componentUuid, tableState),
    );

    expect(store.getActions().length).toEqual(2);

    const loadingAction = store.getActions()[0];
    expect(loadingAction.type).toEqual(GENES_PROPERTIES_LOADING);
    expect(loadingAction).toMatchSnapshot();

    const loadedAction = store.getActions()[1];
    expect(loadedAction.type).toEqual(GENES_PROPERTIES_ERROR);
    expect(loadedAction).toMatchSnapshot();
  });
});
