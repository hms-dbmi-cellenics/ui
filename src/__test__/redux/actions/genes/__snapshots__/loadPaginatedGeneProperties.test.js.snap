// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`loadPaginatedGeneProperties action Dispatches appropriately on error condition 1`] = `
{
  "payload": {
    "componentUuid": "asd",
    "properties": [
      "a",
      "b",
      "c",
    ],
  },
  "type": "genes/propertiesLoading",
}
`;

exports[`loadPaginatedGeneProperties action Dispatches appropriately on error condition 2`] = `
{
  "payload": {
    "componentUuid": "asd",
    "error": [Error: random error!],
    "experimentId": "1234",
    "properties": [
      "a",
      "b",
      "c",
    ],
  },
  "type": "genes/propertiesError",
}
`;

exports[`loadPaginatedGeneProperties action Dispatches appropriately on success condition 1`] = `
{
  "payload": {
    "componentUuid": "asd",
    "properties": [
      "a",
      "b",
      "c",
    ],
  },
  "type": "genes/propertiesLoading",
}
`;

exports[`loadPaginatedGeneProperties action Dispatches appropriately on success condition 2`] = `
{
  "payload": {
    "componentUuid": "asd",
    "data": {
      "a": {
        "dispersions": 1,
      },
      "b": {
        "dispersions": 1,
      },
    },
    "experimentId": "1234",
    "properties": [
      "a",
      "b",
      "c",
    ],
    "total": 2,
  },
  "type": "genes/propertiesLoadedPaginated",
}
`;

exports[`loadPaginatedGeneProperties action Dispatches appropriately on success condition 3`] = `
[
  "1234",
  {
    "limit": 20,
    "name": "ListGenes",
    "offset": 20,
    "orderBy": "gene_names",
    "orderDirection": "ASC",
    "selectFields": [
      "gene_names",
      "a",
      "b",
      "c",
    ],
  },
  [Function],
  [Function],
  {
    "timeout": 60,
  },
]
`;
